### DISCUSSION TITLE
Compound Component Pattern이란?

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#119] [React]
## Compound Component Pattern 이란?

컴파운드 컴포넌트는 두개 이상의 협력하는 컴포넌트들을 서로 유연하게 확장 가능하게 만드는 패턴이다. 컴파운드 컴포넌트 패턴은 기능을 내부에서 관리하고 렌더링에 대한 제어 관한을 구현장소에 위임하는 패턴이다.  이 패턴을 설명할 때 가장 많이 나오는 예시가 바로 html 기본 태그인 select와 option이다. 현재 select를 사용하는 것은 다음과 같다.

```html
<select>
  <option value="value1">key1</option>
  <option value="value2">key2</option>
  <option value="value3">key3</option>
</select>
```

 만약에 리액트에서와 같이 props로 이를 관리하게 된다면 어떻게 될까?
```html
<select option1={value:value1, key1:key} ...>
```
prop에 의존하다보니 한눈에 봐도 해당 select를 재활용하기 어렵다. 또 ui에 대한 변동이 늘어나면 늘어날수록 이에 대한 대처가 어려워지고 prop이 늘어나게 되어 해당 엘리먼트를 재활용하기가 어려워진다. 따라서 option에 대한 스타일이나 로직들을 사용하는 사용처에서 조절할 수 있게 만들어졌다.   Compound Component Pattern은 이러한 유연성에 집중을 하여 만들어졌다.

### Compound Component Pattern의 장점

- UI 변경에 유연하다.
- 상태로직들을 재활용할 수 있다. 공통으로 사용되는 코드들을 응집도 있게 관리할 수 있다.
- 내부 구현을 캡슐화할 수 있다.

### 적용 후기
 해당 패턴을 dropdown을 구성하는데 활용하였다. 현재 dropdown이 활용되는 상황은 아래와 같다.
<img width="153" alt="스크린샷 2022-10-17 오전 12 03 31" src="https://user-images.githubusercontent.com/65863017/196042798-fabb49bc-69fa-4188-97b7-33cb8fb23374.png">
<img width="201" alt="image" src="https://user-images.githubusercontent.com/65863017/196043069-e73fb386-2303-4b83-9c25-0f7844df6b8c.png">

dropdown의 경우 하위 요소들이 펼쳐지는 trigger 버튼과 하위 요소들을 나타내는 Item들이 존재하는 공통 로직을 가지고 있었다. 하지만 하위 요소의 디자인은 세밀하게 달랐다. 따라서 해당 컴포넌트에 Trigger와 Item 컴포넌트를 Dropdown 컴포넌트로 엮어주고 보여줄지 말지를 결정하는 상태를 해당 Dropdown내의 Context로 공유하게 하여 해당 컴포넌트를 구현하였다.

 Compound Component Pattern을 활용해본 결과 상태로직을 재활용하고 UI를 사용처에서 자유롭게 조절이 가능하게 하여 컴포넌트의 재활용성을 높일 수 있었다. 하지만 결국엔 UI의 스타일링을 사용처에 위임하는 것이므로 사용처가 복잡해진다는 단점이 있었다.

**Comment by al-bur** [comment]
왜 뒤를 돌아보는 개발자인가?

**Reply by al-bur** [reply]
최근에도 되돌아봤는지?

**Reply by al-bur** [reply]
최근의 길은 어디 길인지?

**Reply by al-bur** [reply]
회사에 와서 너무 바쁘다. 뒤를 어떻게 돌아보실 건지?

**Comment by al-bur** [comment]
프로젝트에 compound component pattern을 사용한 이유는?

**Reply by byhhh2** [reply]
- 단점은?

**Comment by byhhh2** [comment]
compound component pattern 같이 컴포넌트를 props로 넘겨주는 패턴은 상위 컴포넌트가 좀 더 복잡해지지 않나,
이런 것을 해결한 경험은?

**Comment by al-bur** [comment]
form 요소들이 많을 때는 어떻게 상태관리를 해줄 것인가?

**Comment by byhhh2** [comment]
react는 왜 기본으로 컴포넌트를 memo로 감싸지 않을까?

**Reply by byhhh2** [reply]
- 성능 상의 문제를 어떻게 측정할 수 있을까?

**Reply by byhhh2** [reply]
- lighthouse의 어떤 점수로 확인할 수 있을까요?

**Comment by byhhh2** [comment]
자신만의 컴포넌트 만드는 전략은?

**Reply by byhhh2** [reply]
- 만약 컴포넌트가 복잡해지면 어떻게 분리하시나요?

**Reply by al-bur** [reply]
- 어떻게 페어와 컴포넌트 분리 전략을 상의하였는지?

**Reply by al-bur** [reply]
- 컴포넌트가 비대해지면, 어떻게 컴포넌트를 분리하는 지?

**Comment by byhhh2** [comment]
다음번에도 compound component pattern 쓰실건가요?

**Reply by byhhh2** [reply]
- context API를 어떻게 활용하나요?
